lib/events/dynamoevents/dynamoevents.go

```diff
@@ -348,57 +351,74 @@ func (l *Log) migrateRFD24WithRetry(ctx context.Context) error {
-		if err != nil {
-			return trace.Wrap(err)
-		}
-	}
-
 	// Kick off a background task to migrate events, this task is safely
 	// interruptible without breaking data and can be resumed from the
 	// last point so no checkpointing state needs to be managed here.
 	go func() {
 		for {
-			// Migrate events to the new format so that the V2 index can use them.
-			log.Info("Starting event migration to v6.2 format")
-			err := l.migrateDateAttribute(ctx)
-			if err != nil {
-				log.WithError(err).Error("Encountered error migrating events to v6.2 format")
-			} else {
-				// Remove the old index, marking migration as complete
-				log.Info("Removing old DynamoDB index")
-				err = l.removeV1GSI()
+			err := func() error {
+				hasIndexV1, err := l.indexExists(l.Tablename, indexTimeSearch)
+				if err != nil {
+					return trace.Wrap(err)
+				}
+
+				// Table is already up to date.
+				// We use the existence of the V1 index as a completion flag
+				// for migration. We remove it at the end of the migration which
+				// means it is finished if it doesn't exist.
+				if !hasIndexV1 {
+					return nil
+				}
+
+				// Creates the v2 index if it doesn't already exist.
+				// This function takes a lock internally to prevent race conditions.
+				err = l.createV2GSI(ctx)
+				if err != nil {
+					return trace.Wrap(err)
+				}
+
+				// Acquire a lock so that only one auth server attempts to perform the migration at any given time.
+				// If an auth server does in a HA-setup the other auth servers will pick up the migration automatically.
+				err = backend.RunWhileLocked(ctx, l.backend, rfd24MigrationLock, rfd24MigrationLockTTL, func(ctx context.Context) error {
+					hasIndexV1, err := l.indexExists(l.Tablename, indexTimeSearch)
+					if err != nil {
+						return trace.Wrap(err)
+					}
+
+					if !hasIndexV1 {
+						return nil
+					}
+
+					// Migrate events to the new format so that the V2 index can use them.
+					log.Info("Starting event migration to v6.2 format")
+					err = l.migrateDateAttribute(ctx)
+					if err != nil {
+						return trace.WrapWithMessage(err, "Encountered error migrating events to v6.2 format")
+					}
+
+					// Remove the old index, marking migration as complete
+					log.Info("Removing old DynamoDB index")
+					err = l.removeV1GSI()
+					if err != nil {
+						return trace.WrapWithMessage(err, "Migrated all events to v6.2 format successfully but failed to remove old index.")
+					}
+
+					return nil
+				})
+
 				if err != nil {
-					log.WithError(err).Error("Migrated all events to v6.2 format successfully but failed to remove old index.")
-				} else {
-					break
+					return trace.Wrap(err)
 				}
+
+				return nil
+			}()
+
+			if err == nil {
+				break
 			}
 
 			delay := utils.HalfJitter(time.Minute)
-			log.Errorf("Background migration task failed, retrying in %f seconds", delay.Seconds())
+			log.WithError(err).Errorf("Background migration task failed, retrying in %f seconds", delay.Seconds())
```

Comment:
(Not related to this PR) three lines below we have: 
https://github.com/gravitational/teleport/blob/1807674df717a431f6e9f9f840d267daf96ef89c/lib/events/dynamoevents/dynamoevents.go#L425
```go
select {
   case <-time.After(delay):
   case <-ctx.Done():
	log.WithError(ctx.Err()).Error("Background migration task cancelled")
}
```
Seems that return statment is missing from  `case <-ctx.Done()` and we will reiterate the loop.

Reply:
Good catch, fixed